### PR TITLE
[#11572] Notification feature - Add tests PUT and DELETE route

### PR DESCRIPTION
**Pull request by NicolasCwy**
Part of #11572

**Outline of Solution**
- Add tests for PUT and DELETE routes
- Add missing test annotation for `CreateNotificationAction` and fix test

### CODE REVIEW

File: src/test/java/teammates/ui/webapi/DeleteNotificationActionTest.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+        verifyHttpParameterFailure(requestParams);
+
+        ______TS("Deleting non-existent notification");
+        // Deleting a non-existent notificiation will not throw an error and fail silently
+        requestParams = new String[] {
+                Const.ParamsNames.NOTIFICATION_ID, "non-existent notification",
+        };
+
+        verifyAbsentInDatabase(testNotificationAttribute);
```
Comment (zhaojj2209) [inline src/test/java/teammates/ui/webapi/DeleteNotificationActionTest.java:56]: Seems that this line is redundant since the absence of the test notification is not really relevant to this particular test.

**Review by zhaojj2209 — COMMENTED** [review]
Good coverage!

File: src/test/java/teammates/ui/webapi/UpdateNotificationActionTest.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+        ______TS("Typical Case: Update notification successfully");
+        req = getTypicalUpdateRequest();
+        UpdateNotificationAction action = getAction(req, requestParams);
+        NotificationData res = (NotificationData) action.execute().getOutput();
+
+        NotificationAttributes updatedNotification = logic.getNotification(res.getNotificationId());
+
+        assertEquals(res.getStartTimestamp(), updatedNotification.getStartTime().toEpochMilli());
+        assertEquals(res.getEndTimestamp(), updatedNotification.getEndTime().toEpochMilli());
```
Comment (fsgmhoward) [inline src/test/java/teammates/ui/webapi/UpdateNotificationActionTest.java:57]: I feel these two lines are not consistent with the 4 assertions below.

`style`, `targetUser`... are values input and are "expected values". However, `res` and `updatedNotification` are both fetched from db/backend so they are "actual values". Not saying comparing the two "actual values" are wrong as they are fetched differently and it does verify that the returned value from API is the updated value in DB. However, to be consistent with the below assertions, comparison with `req.getStartTimestamp()` with `updatedNotification` is more preferred in my opinion.

**Review by fsgmhoward — CHANGES_REQUESTED** [review]
LGTM! Just minor nits.

File: src/test/java/teammates/ui/webapi/DeleteNotificationActionTest.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+        DeleteNotificationAction action = getAction(requestParams);
+        JsonResult response = getJsonResult(action);
+        MessageOutput msg = (MessageOutput) response.getOutput();
+        assertEquals("Notification has been deleted.", msg.getMessage());
+
+        verifyAbsentInDatabase(testNotificationAttribute);
+
+        ______TS("Notification ID cannot be null");
+        requestParams = new String[] {
```
Comment (jianhandev) [inline src/test/java/teammates/ui/webapi/DeleteNotificationActionTest.java:44]: Another possible case is when no request param is provided at all

File: src/test/java/teammates/ui/webapi/UpdateNotificationActionTest.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    @Override
+    String getRequestMethod() {
+        return PUT;
+    }
+
+    @Test
+    @Override
+    protected void testExecute() throws Exception {
+        String[] requestParams = new String[] {
+                Const.ParamsNames.NOTIFICATION_ID, testNotificationAttribute.getNotificationId(),
```
Comment (jianhandev) [inline src/test/java/teammates/ui/webapi/UpdateNotificationActionTest.java:37]: Another possible case is when no request param is provided at all

**Review by jianhandev — CHANGES_REQUESTED** [review]

File: src/test/java/teammates/ui/webapi/DeleteNotificationActionTest.java (hunk 1 above)
Comment (NicolasCwy) [reply src/test/java/teammates/ui/webapi/DeleteNotificationActionTest.java:56]: I wanted to ensure that the notification is not in the data base, to be certain that deleting a non-existent notification would cause a silent failure. Though it is unlikely that someone would add a notification with ID `non-existent notification` to the test bundle

Comment (zhaojj2209) [reply src/test/java/teammates/ui/webapi/DeleteNotificationActionTest.java:56]: If I'm not wrong, `testNotificationAttribute` has `notification1` as the ID and not `non-existent notification`, so it seems that `non-existent notification` should be tested here instead?

Comment (NicolasCwy) [reply src/test/java/teammates/ui/webapi/DeleteNotificationActionTest.java:56]: I see what you mean now. Will verify that a notification with ID `non-existent notification` is absent from the database

File: src/test/java/teammates/ui/webapi/UpdateNotificationActionTest.java (hunk 2 above)
Comment (NicolasCwy) [reply src/test/java/teammates/ui/webapi/UpdateNotificationActionTest.java:57]: Have added the changes to my latest commit

**Review by zhaojj2209 — APPROVED** [review]
LGTM!

File: src/test/java/teammates/ui/webapi/UpdateNotificationActionTest.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        // Verify that correctly updated in the DB
+        assertEquals(req.getStartTimestamp(), updatedNotification.getStartTime().toEpochMilli());
+        assertEquals(req.getEndTimestamp(), updatedNotification.getEndTime().toEpochMilli());
+        assertEquals(style, updatedNotification.getStyle());
+        assertEquals(targetUser, updatedNotification.getTargetUser());
+        assertEquals(title, updatedNotification.getTitle());
+        assertEquals(message, updatedNotification.getMessage());
+
+        ______TS("Parameters cannot be null");
+        req = getTypicalUpdateRequest();
```
Comment (jianhandev) [inline src/test/java/teammates/ui/webapi/UpdateNotificationActionTest.java:65]: Sorry for not picking this up this earlier, actually I think we don't need to test null params here because the validation by `getAndValidateRequestBody` will guarantee they are non-null.

File: src/test/java/teammates/ui/webapi/UpdateNotificationActionTest.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        ex = verifyHttpRequestBodyFailure(req, requestParams);
+        assertEquals("Notification title cannot be null", ex.getMessage());
+
+        req = getTypicalUpdateRequest();
+        req.setMessage(null);
+        ex = verifyHttpRequestBodyFailure(req, requestParams);
+        assertEquals("Notification message cannot be null", ex.getMessage());
+
+        ______TS("Timestamps should be greater than 0");
+        req = getTypicalUpdateRequest();
```
Comment (jianhandev) [inline src/test/java/teammates/ui/webapi/UpdateNotificationActionTest.java:86]: Same as null params, I think it is safe to assume that the timestamp will be > 0, if not `getAndValidateRequestBody` itself would throw `InvalidHttpRequestBodyException`. However it will be useful to test that start time should be before end time.

**Review by jianhandev — CHANGES_REQUESTED** [review]

File: src/test/java/teammates/ui/webapi/UpdateNotificationActionTest.java (hunk 5 above)
Comment (NicolasCwy) [reply src/test/java/teammates/ui/webapi/UpdateNotificationActionTest.java:64]: I see, but should we still keep this in for black box testing?

File: src/test/java/teammates/ui/webapi/UpdateNotificationActionTest.java (hunk 6 above)
Comment (NicolasCwy) [reply src/test/java/teammates/ui/webapi/UpdateNotificationActionTest.java:85]: Can add this test case as well, but same as null params, should the existing test be removed?

File: src/test/java/teammates/ui/webapi/UpdateNotificationActionTest.java (hunk 5 above)
Comment (jianhandev) [reply src/test/java/teammates/ui/webapi/UpdateNotificationActionTest.java:64]: Technically there's no harm having those tests, let's keep it first then. Although I certainly think that they are not the most important tests to have.

**Review by ziqing26 — APPROVED** [review]
LGTM!

**Review by t-cheepeng — APPROVED** [review]
LGTM

File: src/test/java/teammates/ui/webapi/UpdateNotificationActionTest.java (hunk 5 above)
Comment (NicolasCwy) [reply src/test/java/teammates/ui/webapi/UpdateNotificationActionTest.java:64]: Sure, one other reason is that the `UpdateNotificationRequest` is not tested, doesn't look like requests themselves are tested directly, referencing the others

**Review by jianhandev — APPROVED** [review]
